pkg/controller/azvolume.go

```diff
@@ -286,31 +285,29 @@ func (r *ReconcileAzVolume) triggerDelete(ctx context.Context, azVolume *azdiskv
 
 				// wait async and report error to go channel
 				wg.Add(1)
-				go func(ctx context.Context, waiter *watcher.ConditionWaiter, i int) {
+				go func(ctx context.Context, waiter *watcher.ConditionWaiter, attachment azdiskv1beta2.AzVolumeAttachment) {
 					defer waiter.Close()
 					defer wg.Done()
 					_, err := waiter.Wait(ctx)
 					if err != nil {
-						errors[i] = err
-						atomic.AddUint32(&numErrors, 1)
+						errorMessageCh <- fmt.Sprintf("%s: %v", attachment.Name, err)
 					}
-				}(deleteCtx, waiter, i)
+				}(deleteCtx, waiter, attachment)
 			}
 
 			// start waiting for the attachment clean up to complete (outside of the (if err == nil) to avoid child workflow finishing after parent workflow finishing)
 			wg.Wait()
+			close(errorMessageCh)
```

Comment:
`errorMessageCh` was not created as a buffered channel, so the goroutines would never be able to exit, as it would be waiting for a recipient. That said, `wg.Wait()` would be blocked indefinitely if an error occurs.

So, you would either need to:
* make the `errorMessageCh` a buffered channel of `size = len(attachments)`
OR
* get rid of waitgroup.
* send result to `errorMessageCh` in both success and failure cases (`nil` value would be sent in case of success, and `non-nil` error would be sent in case of failure`
* loop over `len(azVolumeAttachment` and collect `non-nil` errors.
* close channel once completed looping.

Reply:
for the buffered `errorMessageCh` case, we would need to create another goroutine to wait and close the channel as follows: 
```go func() { wg.Wait() close(errors ) }()```
and after that we can loop over `errorMessageCh` to get all the errors: 
```for error := range errors { // update status }```

Reply:
Done. I prefer the second one since it gets rid of the waitgroup which is more straightforward.